layers/gpu/spirv/non_bindless_oob_texel_buffer_pass.cpp

```diff
@@ -0,0 +1,234 @@
+#include "module.h"
+#include <spirv/unified1/spirv.hpp>
+#include <iostream>
+
+#include "generated/instrumentation_non_bindless_oob_texel_buffer_comp.h"
+#include "gpu/shaders/gpu_shaders_constants.h"
+
+namespace gpu {
+namespace spirv {
+
+NonBindlessOOBTexelBufferPass::NonBindlessOOBTexelBufferPass(Module& module) : Pass(module) { module.use_bda_ = true; }
+
+static LinkInfo link_info = {instrumentation_non_bindless_oob_texel_buffer_comp,
+                             instrumentation_non_bindless_oob_texel_buffer_comp_size,
+                             LinkFunctions::inst_non_bindless_oob_texel_buffer, 0, "inst_non_bindless_oob_texel_buffer"};
+
+// By appending the LinkInfo, it will attempt at linking stage to add the function.
+uint32_t NonBindlessOOBTexelBufferPass::GetLinkFunctionId() {
+    if (link_function_id == 0) {
+        link_function_id = module_.TakeNextId();
+        link_info.function_id = link_function_id;
+        module_.link_info_.push_back(link_info);
+    }
+    return link_function_id;
+}
+
+uint32_t NonBindlessOOBTexelBufferPass::CreateFunctionCall(BasicBlock& block, InstructionIt* inst_it,
+                                                           const InjectionData& injection_data) {
+    assert(access_chain_inst_ && var_inst_);
+    const Constant& set_constant = module_.type_manager_.GetConstantUInt32(descriptor_set_);
+    const Constant& binding_constant = module_.type_manager_.GetConstantUInt32(descriptor_binding_);
+    const uint32_t descriptor_index_id = CastToUint32(descriptor_index_id_, block, inst_it);  // might be int32
+
+    const uint32_t opcode = target_instruction_->Opcode();
+    const uint32_t image_operand_position = OpcodeImageOperandsPosition(opcode);
+    if (target_instruction_->Length() > image_operand_position) {
+        const uint32_t image_operand_word = target_instruction_->Word(image_operand_position);
+        if ((image_operand_word & (spv::ImageOperandsConstOffsetMask | spv::ImageOperandsOffsetMask)) != 0) {
+            // TODO - Add support if there are image operands (like offset)
+        }
+    }
+
+    // Use the imageFetch() parameter to decide the offset
+    // TODO - This assumes no depth/arrayed/ms from AnalyzeInstruction
+    descriptor_offset_id_ = CastToUint32(target_instruction_->Operand(1), block, inst_it);
+
+    const uint32_t function_result = module_.TakeNextId();
+    const uint32_t function_def = GetLinkFunctionId();
+    const uint32_t bool_type = module_.type_manager_.GetTypeBool().Id();
+
+    block.CreateInstruction(
+        spv::OpFunctionCall,
+        {bool_type, function_result, function_def, injection_data.inst_position_id, injection_data.stage_info_id,
+         descriptor_array_size_id_, set_constant.Id(), binding_constant.Id(), descriptor_index_id, descriptor_offset_id_},
+        inst_it);
+
+    return function_result;
+}
+
+void NonBindlessOOBTexelBufferPass::Reset() {
+    access_chain_inst_ = nullptr;
```

Comment:
I think you could go:
```suggestion
*this = NonBindlessOOBTexelBufferPass();
```
Simpler and makes sure no member is forgotten.
But not 100% sure as `NonBindlessOOBTexelBufferPass` is inheriting from `Pass`

Reply:
this would reset `link_function_id` back to `0`, that is suppose to be the same between runs... agree this is error prone, but also the way around it would be to have a struct in the class that could be reset like this

Reply:
will add to a future enhancement todo list